lightning/src/routing/gossip.rs

```diff
@@ -1555,12 +1603,40 @@ impl<L: Deref> NetworkGraph<L> where L::Target: Logger {
 		}
 	}
 
-	/// Marks a node in the graph as failed.
-	pub fn node_failed(&self, _node_id: &PublicKey, is_permanent: bool) {
-		if is_permanent {
-			// TODO: Wholly remove the node
-		} else {
-			// TODO: downgrade the node
+	/// Marks a node in the graph as permanently failed, effectively removing it and its channels
+	/// from local storage.
+	///
+	/// If called with the "std" feature enabled, the removed node and channels will also be tracked
+	/// as removed so they're not re-added on gossip re-broadcast.
+	pub fn node_failed_permanent(&self, node_id: &PublicKey) {
+		let node_id = NodeId::from_pubkey(node_id);
+		let mut channels = self.channels.write().unwrap();
+		let mut nodes = self.nodes.write().unwrap();
+		#[cfg(feature = "std")]
+		let mut removed_channels = self.removed_channels.lock().unwrap();
+		#[cfg(feature = "std")]
+		let mut removed_nodes = self.removed_nodes.lock().unwrap();
+		#[cfg(feature = "std")]
+		let time = Instant::now();
+
+		if let Some(node) = nodes.remove(&node_id) {
+			for scid in node.channels.iter() {
+				if let Some(chan_info) = channels.remove(scid) {
+					let other_node_id = if node_id == chan_info.node_one { chan_info.node_two } else { chan_info.node_one };
+					if let BtreeEntry::Occupied(mut node_entry) = nodes.entry(other_node_id) {
+						node_entry.get_mut().channels.retain(|chan_id| {
+							*scid != *chan_id
+						});
+						if node_entry.get().channels.is_empty() {
+							node_entry.remove_entry();
+						}
+					}
+					#[cfg(feature = "std")]
+					removed_channels.insert(*scid, time);
+				}
+			}
+			#[cfg(feature = "std")]
+			removed_nodes.insert(node_id, time);
```

Comment:
If the other node has no more channels, should it also be removed? Or will that be a problem if another channel is advertised for it?

Reply:
Ah, actually, I guess this only for permanent failures for the passed in node.

Reply:
Yeah exactly. We don't want to track it as removed as it wasn't a permanent failure.